tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15358,10 +15358,46 @@ void DiagnoseNodeEntry(Sema &S, FunctionDecl *FD, HLSLShaderAttr *Attr) {
   return;
 }
 
+void TryAddShaderAttrFromTargetProfile(Sema &S, FunctionDecl *FD) {
+  const std::string &EntryPointName = S.getLangOpts().HLSLEntryFunction;
+
+  // if this is the Entry FD, then add the target profile
+  // shader attribute to the FD and carry on with validation
+  // otherwise, with no shader attribute, just return
+  if (EntryPointName.empty()) {
+    return;
+  }
+
+  // if this FD isn't the entry point, then there's no
+  // shader attribute to work with, so just return
+  if (EntryPointName != FD->getIdentifier()->getName() || 
+    (EntryPointName == FD->getIdentifier()->getName() && !FD->isGlobal())) {
+    return;
+  }
+
+  std::string profile = S.getLangOpts().HLSLProfile;
+  const ShaderModel *SM = hlsl::ShaderModel::GetByName(profile.c_str());
+  const llvm::StringRef fullName = ShaderModel::FullNameFromKind(SM->GetKind());
+
+  // don't add the attribute for an invalid profile, like library
+  if (fullName.empty()) {
+    return;
+  }
+
+  HLSLShaderAttr *pShaderAttr =
```

Comment:
We should also check if there's an existing `shader` attribute, and if so, skip adding it after making sure it matches the `fullName` for the shader stage.

Reply:
If I'm comparing equality for target profile and any existing attributes, it begs the question of whether this function should be responsible for emitting a diagnostic if there's a mismatch. Instead of leaving it to later validation, should I emit a diagnostic if, say, the "pixel" attr is defined on the decl, but the target profile is "vertex", and delete that validation code when it happens later?

Reply:
If we have specific diagnostics elsewhere for trying to compile an entry point marked with a conflicting shader attribute later, then we can probably skip both the diagnostic and the addition of the conflicting attribute here.